Python 3.x support in nilearn

bcipolli opened the pull request:
@lesteve has helped finish the work here, and I removed the dependency on #371, so I've removed the "WIP" and I think things are more ready to go.

All code, tests, and examples should be runnable in Python 3.x.  We've enabled test coverage for Python 3.x, and the build is passing.  In addition, I tested with a version on #371 (resolving warnings), and verified that no new warnings are introduced by this code.

The following conventions were used to migrate `nilearn` code to Python 2/3 compatibility:
* All `nilearn` imports should be relative (except, perhaps, in tests)
* Write `except Exception, e` as `except Exception as e`
* Write `isinstance(x, types.ListType)` as `isinstance(x, list)`
* Write `callable(x)` as `isinstance(x, collections.Callable)`
* Write `operator.isSequenceType(x)` as `isinstance(x, collections.Sequence)`
* Write `operator.isNumberType(x)` as `isinstance(x, collections.Number)`
* Write `nose.tools.assert_equals` as `nose.tools.assert_equal`
* Write `nilearn._utils.testing.assert_raises_regexp` as `nilearn._utils.testing.assert_raises_regex` (mirrors nose clarification)
* Write `<Iterable>.next()` as `next(<Iterable>)`
* Write `filename = os.tempnam()` as `(_, filename) = tempfile.mkstemp()`
* Write `not key in dict` as `key not in dict` (this actually warns in Python 3.x!)
* Make sure to use `with open('file') as f' syntax; any file pointers that are not used in a context manager nor explicitly closed throw warnings.

**Python 3 changes that require care to be backwards compatible**
* Non-unicode strings must be prefixed with `b`: `b'string'`(mostly an issue interacting with numpy, but also some low-level operations like gzip code)
* Write `print "%s" % (message,)` as `print("%s" % (message,))`
* Write integer division `3 / 2 == 1` as `3 // 2 == 1`
* Write `<dict>.iteritems()` as `<dict>.items()` (there is no iteritems in Python 3)
* Write `<dict>.itervalues()` as `<dict>.values()` (there is no itervalues in Python 3)
* Write `itertools.izip()` as `zip()` (there is no `izip` in Python 3)
* Write octal string `0100` `0o100`
* Write `round` as `np.round` (avoids a numpy issue in Python 3)

**Importing things from `six` instead of places you're used to**
* Write `basestring` as `six.string_types` (requires `import six`)
* Use the new urllib object structure instead of ulrllib/urllib2
    * import from `six.moves.urllib`.
    * there are no base-level functions/objects, all are present in subpackages
* Import  `StringIO` and `cPickle from `six.moves`

**Changes in how you design / write code**
* Think carefully about any use of `xrange`.  If you *require* an iterator, then you must import `past.builtins.xrange`.  Otherwise, just use `range` (which is a list in Python 2, and an interator in Python 3).
* Most things that used to generate lists now generate iterators; any time you need a list, make sure to wrap it in `list()`
    * **THIS WILL GET YOU!**
    * Includes:
        * `dict.keys()`, `dict.values()`
        * `zip`
        * `range`
        * `map`

lesteve commented:
First thanks a lot for removing the warnings related commits from the python 3 ones!

Here are a few general comments, before I look at the changes in more details:

- [ ] we want to either use six or future but not both. Personally I'd be in favor of the latter as it l
- [ ] we should leave the documentation generation for a separate PR. Please revert any changes to doc/sphintext.

Also this is quite a long PR to review. That would be really great if you could keep unrelated commits, like "shorten numpy version idiom" for separate PRs.

> All nilearn imports should be relative (except, perhaps, in tests)

not perhaps, tests should use absolute imports. For example it allows you to run a test from your git checkout with any nilearn that happens to be in your PYTHONPATH. For example this is quite useful to make sure that a test you just added was actually failing with a previous released nilearn version.

lesteve commented on README.rst:
```diff
@@ -40,7 +40,8 @@ The required dependencies to use the software are:
 * Scikit-learn >= 0.12.1
 * Nibabel >= 1.1.0.
-
+* Six
```
please remove the newline before and add a newline after (the latter is to make the README.rst render properly on PyPI)

lesteve commented on README.rst:
```diff
@@ -40,7 +40,8 @@ The required dependencies to use the software are:
 * Nibabel >= 1.1.0.
-
+* Six
+=======
```
Not too sure where this === comes from but I think it should be removed unless there is a missing title in the rst.

lesteve commented on doc/issues.txt:
```diff
@@ -0,0 +1,18 @@
+Create RSA code
```
you probably didn't mean to include this file

lesteve commented on examples/decoding/plot_haxby_full_analysis.py:
```diff
@@ -23,10 +23,12 @@
 import numpy as np
-labels = np.recfromcsv(haxby_dataset.session_target[0], delimiter=" ")
+labels = np.recfromcsv(haxby_dataset.session_target[0], delimiter=" ",
+                       dtype=[('labels', 'U12'), ('chunks', '<i8')])
```
I think dtype doesn't need to be mentioned here similarly to what is done in the other haxby examples.

bcipolli replied:
Thanks for catching this.  I think this was my attempted solution for examples, before simply prefixing all strings as byte-strings.  I've removed now.

lesteve commented on nilearn/_utils/testing.py:
```diff
@@ -24,11 +28,11 @@
 
 try:
-    from nose.tools import assert_raises_regexp
+    from nose.tools import assert_raises_regex
```
slight niggle: assert_raises_regex does not exist for python 2.7 because it is called assert_raises_regexp (with a 'p' at the end) we should do something like this:

```python
try:
    from nose.tools import assert_raises_regex
except ImportError
    try:
        from nose.tools import assert_raises_regexp as assert_raises_regex
    except ImportError:
        # for Python 2.6
```

Side-comment: very annoyingly PEP8 is not stable with respect to search and replace which means that most of the assert_raises_regexp -> assert_raises_regex replacements in this PR have introduced PEP8 violations...

bcipolli replied:
Fixed. Also dealt with those PEP8 violations... you are correct: that IS very annoying :)

lesteve commented on nilearn/_utils/testing.py:
```diff
@@ -56,6 +60,34 @@ def assert_warns(warning_class, func, *args, **kw):
         return output
 
 
+def assert_warns_regex(warning_class, warning_regex,
```
I may have missed it but it looks like you are not using this function in this PR (kind of makes sense that it belongs in the warnings-related PR)

bcipolli replied:
Nice catch.  There were two more commits from the `fix_warnings` branch that I missed here (from their commit comment, I couldn't tell).  Rebased to remove them.

lesteve commented on nilearn/decoding/searchlight.py:
```diff
@@ -176,7 +177,7 @@ def _group_iter_search_light(list_rows, estimator, X, y,
                     crlf = "\n"
                 percent = float(i) / len(list_rows)
-                percent = round(percent * 100, 2)
+                percent = np.round(percent * 100, 2)
```
My understanding is that the round -> np.round replacement is only needed when the thing you are rounding is a numpy scalar. It doesn't seem necessary here.

bcipolli replied:
My understanding is the same.  I replaced it here while replacing in other places, as it seemed more robust to future changes and a good habit to simply use `np.round` in any code already importing numpy.

Happy to change it back.

lesteve commented on nilearn/input_data/nifti_maps_masker.py:
```diff
@@ -3,6 +3,7 @@
 """
 
 import numpy as np
+from six import string_types
```
not used it looks like

lesteve replied:
> not used it looks like

still seems to be the case

lesteve commented on nilearn/input_data/nifti_labels_masker.py:
```diff
@@ -7,8 +7,6 @@
 from sklearn.base import BaseEstimator, TransformerMixin
 from sklearn.externals.joblib import Memory
 
-import nibabel
```
not really related to the PR ...

lesteve commented on nilearn/mass_univariate/permuted_least_squares.py:
```diff
@@ -5,6 +5,8 @@
 # Author: Benoit Da Mota, <benoit.da_mota@inria.fr>, sept. 2011
 #         Virgile Fritsch, <virgile.fritsch@inria.fr>, jan. 2014
 import warnings
+from past.builtins import xrange
```
not used

lesteve commented on nilearn/plotting/cm.py:
```diff
@@ -157,7 +158,7 @@ def alpha_cmap(color, name='', alpha_min=0.5, alpha_max=1.):
 _cmap_d = dict()
 
-for _cmapname in _cmaps_data.keys():
+for _cmapname in list(_cmaps_data.keys()):
```
just wondering whether you need the list here, since you are just iterating over the keys

bcipolli replied:
Fixed.

lesteve commented on nilearn/region.py:
```diff
@@ -6,6 +6,7 @@
 """
 # Author: Philippe Gervais
 # License: simplified BSD
+from past.builtins import xrange
```
not used

lesteve replied:
xrange being imported but not used seems to happen quite a few times, can you check all of them?

lesteve commented on nilearn/datasets.py:
```diff
@@ -2497,22 +2498,24 @@ def fetch_abide_pcp(data_dir=None, n_subjects=None, pipeline='cpac',
+    pheno = BytesIO(pheno)
     # We enforce empty comments because it is 'sharp' by default
-    pheno = np.recfromcsv(pheno, comments=[], case_sensitive=True)
+    pheno = np.recfromcsv(pheno, comments='xyzpdq', case_sensitive=True)
```
is this magical 'xyzpdq' really needed?

bcipolli replied:
... fixed all, except this one.  I am investigating this, I found this to be necessary (but will double-check).

lesteve commented on nilearn/tests/test_masking.py:
```diff
@@ -16,11 +16,10 @@
                              compute_background_mask, unmask, _unmask_3d,
                              _unmask_4d, intersect_masks, MaskWarning)
+from nilearn._utils.testing import (write_tmp_imgs, assert_raises_regex,
+                                    assert_warns_regex)
```
assert_warns_regex isn not used

bcipolli replied:
Thanks for the catch!  This made me realize, just now: one downside of using parentheses in imports is that sublime text no longer indicates which imports are being used.

lesteve commented on nilearn/tests/test_masking.py:
```diff
@@ -16,11 +16,10 @@
-
-np_version = (np.version.full_version if hasattr(np.version, 'full_version')
-              else np.version.short_version)
+np_version = getattr(np.version, 'full_version', np.version.short_version)
```
Interestingly enough this line is actually overriden by the next one which redefines np_version ...

bcipolli replied:
I'll revert this back to untouched; I thought it was causing an error from a merge conflict, but turns out it was in an import above.

lesteve commented on nilearn/group_sparse_covariance.py:
```diff
@@ -9,6 +9,7 @@
 import collections
 import operator
 import itertools
+from past.builtins import xrange
```
not used

bcipolli replied:
I see what happened; this was caused by rebase / merge / rebase confusion.  I eliminated the use of past completely now; none were actually meaningful.

lesteve commented on nilearn/tests/test_signal.py:
```diff
@@ -329,7 +328,7 @@ def test_clean_confounds():
     # C and F order might take different paths in the function. Check that the
     # result is identical.
-    n_features = 1001
+    n_features = 1000
```
I seem to remember this was related to removing one warning so this change and the one above can be removed

bcipolli replied:
Missed this commit.  I've rebased to remove it, thanks!

lesteve commented on nilearn/version.py:
```diff
@@ -28,9 +28,22 @@
         'required_at_installation': False}),)
 
 
+def get_min_version(module_name):
```
does not seem to be used

bcipolli commented:
Thanks @lesteve for the comments!  Looks like things got a little messy after all from the merge / demerge / remerge I did; I will also review the code manually in addition to addressing your comments.

Re: the dependency on `six`, did you have any ideas o removing the dependency on `six.string_types`?  I guess I could just copy their code into a `_utils` function, but that might be confusing for developers? (e.g.  `from nilearn._utils.compat import string_types`)?

lesteve commented:
> Re: the dependency on six, did you have any ideas o removing the dependency on six.string_types

My guess was that future has something similar but I have to say I haven't checked.

bcipolli commented:
@lesteve I didn't find anything, and I searched for a bit.  The code in `six` isn't complicated to copy, I just wanted to check that it would be an acceptable solution.

lesteve commented:
What about http://python-future.org/compatible_idioms.html#basestring?

bcipolli commented:
>What about http://python-future.org/compatible_idioms.html#basestring?
The first option adds a dependency to `past`, the second doesn't consider byte strings as strings.

bcipolli commented:
As it turns our, `sklearn.externals` includes `six`.  How about I migrate to using `sklearn.externals.six`?

bcipolli commented:
@lesteve I made all of the requested changes, I believe.  I chose to remove the dependency on `future` and keep that of `six`, since `six` is much more deeply embedded and is used in `scikit-learn`.  It is now listed as a dependency (which can be removed when minversion of `sklearn` >= 0.14).  Please let me know if that's acceptable.

I will do a line-by-line code review myself later today and update here.

lesteve commented:
>>   What about http://python-future.org/compatible_idioms.html#basestring?

> The first option adds a dependency to past, the second doesn't consider byte strings as strings.

:confused: past is provided by python-future, isn't it?

I may be dead wrong but I think your tests are going to fail because 'from past.builtins import bla' are going to raise, since future is not installed any more by setup.py ...

bcipolli commented:
>:confused: past is provided by python-future, isn't it?

Right, good point!  I decided to stick with `six` because it is heavily used in this PR; moving off of it would be challenging.  In addition, `sklearn` seems to be using `six`.  So that kind of made up things for me to respond to you.  But as always, willing to take a different direction if requested.

>I may be dead wrong but I think your tests are going to fail because 'from past.builtins import bla' are going to raise, since future is not installed any more by setup.py ...

The only dependency on `past` was the imports of `xrange`, and none of those should have been there anyway.  Those are all removed now, so I believe things should run well.  Let's see! :)

lesteve commented:
> Those are all removed now, so I believe things should run well. Let's see! :)

Right should be fine then, I was probably still looking at an old diff.

bcipolli commented:
Just finished a code review, no obvious issues remaining to me.  I just updated the "magic_string" info and migrated to use `xrange` from `six.moves`.  Once the tests pass, I think this is ready for further review.

lesteve commented on doc/sphinxext/gen_rst.py:
```diff
@@ -8,19 +8,18 @@
 Files that generate images should start with 'plot'
```
Looks like there are still have changes in doc/sphintext that you haven't reverted

bcipolli replied:
Very strange, thanks for catching.  In fact none of the changes were present, I wonder if I pulled the commit when rebasing.  Did the same copy from master again!

lesteve commented on examples/connectivity/plot_adhd_covariance.py:
```diff
@@ -26,7 +26,7 @@ def plot_matrices(cov, prec, title):
     # Put zeros on the diagonal, for graph clarity.
     size = prec.shape[0]
-    prec[range(size), range(size)] = 0
+    prec[list(range(size)), list(range(size))] = 0
```
would

```python
diagonal_indices = np.diag_indices(prec.shape[0])
 prec[diagonal_indices] = 0
```
work?

bcipolli replied:
Just looked, turns out there's a `np.fill_diagonal` function.  Both it, and `diag_indices` are new in numpy 1.4, and nilearn targets >= 1.3.

lesteve replied:
Arrgh, probably one for soonish when we raise the minimum versions similarly to #397.

bcipolli replied:
How about opening an issue to track line changes for when the version gets bumped?  I'd be happy to do that, if that sounds useful.

lesteve commented on examples/manipulating_visualizing/plot_affine_transformation.py:
```diff
@@ -69,7 +69,7 @@
 
 # Rotate it slightly
-angle = np.pi / 180 * 15
+angle = np.pi / 180. * 15
```
I don't think this is needed since np.pi is a float

lesteve commented on nilearn/_utils/testing.py:
```diff
@@ -238,14 +239,14 @@ def generate_regions_ts(n_features, n_regions,
     boundaries = np.zeros(n_regions + 1)
     boundaries[-1] = n_features
-    boundaries[1:-1] = rand_gen.permutation(range(1, n_features)
+    boundaries[1:-1] = rand_gen.permutation(list(range(1, n_features))
```
could we use np.arange here?

bcipolli commented:
@lesteve Addressed all of the requested issues except the `np.diag_indices` change, due to the numpy version issue.

lesteve commented on nilearn/_utils/testing.py:
```diff
@@ -96,7 +102,9 @@ def write_tmp_imgs(*imgs, **kwargs):
             warnings.simplefilter("ignore", RuntimeWarning)
             for img in imgs:
-                filename = os.tempnam(None, "nilearn_") + ".nii"
+                (_, filename) = tempfile.mkstemp(prefix="nilearn_",
```
slight niggle, the parentheses are not needed

lesteve replied:
i.e. you can use `_, filename = ...`

bcipolli replied:
good point.  since only one argument is being used, how do you feel about removing the output argument list and just subscripting the function call: 
`filename = tempfile.mkstemp(prefix="nilearn_", suffix=".nii", dir=None)[1]`

lesteve replied:
Probably very subjective, but I find it reads better with `_, filename = ...`.

lesteve commented on nilearn/_utils/testing.py:
```diff
@@ -186,14 +186,15 @@ def __call__(self, *args, **kwargs):
             if basename in self.csv_files:
                 array = self.csv_files[basename]
+
                 # np.savetxt does not have a header argument for numpy 1.6
```
Another one to keep in mind when we move the numpy requirement to 1.6. If you have some spare cycles, feel free to open an issue as you mentioned so that we don't forget about these things and reference #397 inside.

lesteve replied:
Actually I just checked and I misread the comment: there is no header argument in np.savetxt for numpy 1.6, which by the way is probably the older version we are testing on travis-ci.

lesteve commented on nilearn/decoding/tests/test_searchlight.py:
```diff
@@ -20,7 +20,7 @@ def test_searchlight():
     mask_img = nibabel.Nifti1Image(mask.astype(np.int), np.eye(4))
     # Create a condition array
-    cond = np.arange(frames, dtype=int) > frames / 2
+    cond = np.arange(frames, dtype=int) > frames / 2.
```
I don't think this is needed since the dtype is int and frames is a multiple of 2 anyway.

lesteve replied:
You can use // if that makes you feel a lot better ...

lesteve commented on nilearn/input_data/nifti_masker.py:
```diff
@@ -199,8 +199,8 @@ def fit(self, imgs=None, y=None):
         self.mask_img_.get_data()
-        if self.verbose > 10:
-            print "[%s.fit] Finished fit" % self.__class__.__name__
+        if self.verbose > 0:
```
I am guessing this change (verbose > 10 rather than verbose > 0) wasn't meant to be there

bcipolli replied:
Sorry, I thought it was previously requested that I change this.  All of the similar prints from within this function were for `verbose > 0`, so I thought I was asked to tweak this one while I was there.

Please confirm either way, thanks!

lesteve replied:
My recollection is that this is a print statement before immediately returning and that you want this only in some edge cases, for which you would set a crazy high verbosity. 

I seem to remember one of such edge cases was if you want to differentiate between the fit taking a long time and the next step in your analysis taking a long time (and not doing any printing when it starts for some stupid reason).

lesteve commented on nilearn/mass_univariate/permuted_least_squares.py:
```diff
@@ -5,8 +5,10 @@
+
 import numpy as np
 from scipy import linalg
+from six.moves import xrange
```
Unneeded changes

lesteve replied:
Ignore this one too ...

lesteve commented on nilearn/region.py:
```diff
@@ -9,6 +9,7 @@
 
 import numpy as np
 from scipy import linalg, ndimage
+from six.moves import xrange
```
Uneeded change

bcipolli replied:
`xrange` is a builtin in Python 2.x, but unavailable in Python 3.x.  Converting to `range` seemed suboptimal, since some of the ranges being iterated over are quite large.  So I added the imports so that the `xrange` function would be available in both.

Do you prefer simply converting all `xrange` to `range` then?

lesteve replied:
Ah sorry for that, I was in automatic diff processing mode and missed that completely, sorry for the noise ! I think it's fine to keep with xrange.

bcipolli replied:
No problem, it's a huge diff that you've gone over a few times... really thanks!

bcipolli commented:
OK, touched the few that were requested. :)

lesteve commented:
> OK, touched the few that were requested. :)

Great stuff! I had something running my computer at work to make sure that all the datasets were downloading cleanly and the examples ran fine with Python 3. I'll double-checked that tomorrow and once this one is merged, we will all have one cheap excuse less to keep using Python 2!

lesteve commented:
No clue why the coveralls is misbehaving like this ...

lesteve commented:
The only thing I could find in the travis [log](https://travis-ci.org/nilearn/nilearn/jobs/53106433) 

```
Coverage.py warning: Trace function changed, measurement is likely wrong: None
```

For reference a travis [log](https://travis-ci.org/nilearn/nilearn/jobs/53044084) from master.

lesteve commented:
OK I am going to merge this one. Thanks a lot @bcipolli for your continued work on this !

I haven't figured out the whole thing but I convinced myself there is something spurious going on with the coverage. For example I manually checked that some line labeled as miseed were actually  executed using 1/0 markers.

A temporary work-around that I found for now: disabling doctests within nose and use the python 3 distribution in Travis.
 
Note that there is some known undesired interactions between coverage and doctest mentioned [here](http://nose.readthedocs.org/en/latest/plugins/doctests.html#source).

I'll keep investigating to find a more statisfactory long-term solution.

bcipolli commented:
Thanks @lesteve for your help on this one!

A note to others that `six` is now a dependency of `nilearn`, so you may need to update your Python install.

Also to those who have outstanding PRs, I recommend restarting the Travis build so that your code can be tested for Python 3.4 compatibility.

GaelVaroquaux commented:
Great job! Thanks a lot
